resource/config_map.go

```diff
@@ -0,0 +1,198 @@
+	}
+	if config.ResourceName == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.ResourceName must not be empty", config)
+	}
+
+	c := &ConfigMap{
+		desiredFn: config.DesiredFn,
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
+
+		name:         config.Name,
+		namespace:    config.Namespace,
+		resourceName: config.ResourceName,
+	}
+
+	return c, nil
+}
+
+func (c *ConfigMap) Name() string {
+	return c.resourceName
+}
+
+func (c *ConfigMap) EnsureCreated(ctx context.Context, obj interface{}) error {
+	var err error
+
+	printName := newPrintName(c.namespace, c.name)
+
+	var desired *corev1.ConfigMap
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding desired ConfigMap %#q", printName))
+
+		desired, err = c.desiredFn(ctx, obj)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		validateDesiredObject(desired, c.namespace, c.name)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found desired ConfigMap %#q", printName))
+	}
+
+	var current *corev1.ConfigMap
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("finding current ConfigMap %#q", printName))
+
+		current, err = c.k8sClient.CoreV1().ConfigMaps(c.namespace).Get(c.name, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("did not find current ConfigMap %#q", printName))
+
+		} else if err != nil {
+			return microerror.Mask(err)
+
+		} else {
+			c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found current ConfigMap %#q", printName))
+		}
+	}
+
+	if current != nil {
+		c.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("updating ConfigMap %#q", printName))
+
+		toUpdate := newConfigMapToUpdate(current, desired)
+
+		if toUpdate != nil {
+			_, err := c.k8sClient.CoreV1().ConfigMaps(c.namespace).Update(toUpdate)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			c.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("updated ConfigMap %#q", printName))
+
+		} else {
+			c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("ConfigMap %#q is up to date", printName))
+		}
+
+	} else {
+		c.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("creating ConfigMap %#q", printName))
+
+		_, err := c.k8sClient.CoreV1().ConfigMaps(c.namespace).Create(desired)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		c.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("created ConfigMap %#q", printName))
+	}
+
+	return nil
+}
+
+func (c *ConfigMap) EnsureDeleted(ctx context.Context, obj interface{}) error {
+	var err error
+
+	printName := newPrintName(c.namespace, c.name)
+
+	{
+		c.logger.LogCtx(ctx, "level", "info", "message", fmt.Sprintf("deleting ConfigMap %#q", printName))
+
+		err = c.k8sClient.CoreV1().ConfigMaps(c.namespace).Delete(c.name, &metav1.DeleteOptions{})
```

Comment:
Here we could just use the name and namespace from the desired state as described above.

Reply:
I considered that. Look here https://github.com/giantswarm/operatorkit/pull/240/files#r224986690.